README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ROCFTP.MMS

<!-- badges: start -->

<!-- badges: end -->

The goal of ROCFTP.MMS is to generate perfect sample from a given
posterior. It takes a posterior, and generates two Markov chains by
Metropolis with a multishift proposal. This proposal is monotone and
forces these two Markov chains started from the two extreme points of
the most interest range to be coalesced. This mechanism is used in
ROCFTP to generate a perfect sample.

## Installation

You can install the released version of ROCFTP.MMS from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("ROCFTP.MMS")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("nabipoor/ROCFTP.MMS")
```

## Example

Basic examples provided in the Help document.


NEWS.md
This is the first version of this package. There is no news.


tests/testthat.R
library(testthat)
library(ROCFTP.MMS)

test_check("ROCFTP.MMS")


tests/testthat/testinputs.R
library(testthat)
library(ROCFTP.MMS)

expect_equal(ROCFTP.MMS(100,c(20,40), function(x) {dnorm(x, mean=30, sd=1, log=TRUE)}, 0.5, log=FALSE),
             noquote("Error: Change length of coalescence block or sigma or log parameter"))
expect_equal(ROCFTP.MMS("a",c(20,40), function(x) {dnorm(x, mean=30, sd=1, log=TRUE)}, 0.5, log=TRUE),
            noquote("Error: LB should be an integer"))
expect_equal(ROCFTP.MMS(-10,c(20,40), function(x) {dnorm(x, mean=30, sd=1, log=TRUE)}, 0.5, log=TRUE),
             noquote("Error: LB should be an integer"))
expect_equal(ROCFTP.MMS(100,c(20,40), function(x) {dnorm(x, mean=30, sd=1, log=TRUE)}, -0.5, log=TRUE),
             noquote("Error: sigma should be a positive number"))
expect_equal(ROCFTP.MMS(100,c(20,40,5), function(x) {dnorm(x, mean=30, sd=1, log=TRUE)}, 0.5, log=TRUE),
             noquote("Error: The length of start vector should be two"))


R/CFTP.R

#' @include LR.R multishift.R M.MSH.sampler.R
#'
# Function: CFTP
#
# Coupling From The Past generate one block with coalescence
# LB: time of block for coalescing
# start: initial values of two paths (the most interest range)
# post: log of target or posterior distribution
# sigma: standard deviation for multshift proposal
# log: A logical value of T or F which allows to log form of posterior
#
# Authors: Majid Nabipoor, Duncan Murdoch
# revised: Oct. 2021

CFTP<- function (LB,start, post,sigma, log=FALSE)
{
  allresult<- NULL
  y<- NULL
  init<- start
  T<- 0
  r<-0
  repeat{
    samples<-  M.MSH.sampler(LB,init, post, sigma, log)
    T<- T+ nrow(samples)
    r<-r+1
    na<-matrix(c(rep(NA, (T-1)*length(start)),start), nrow=T, byrow=TRUE)
    allresult<- cbind(rbind(allresult, samples),na)
    y<- samples
    init<- allresult[T,]
    if(length(unique(samples[nrow(samples),]))==1 || r>50) break
  }
  ifelse(r<50,return(y),return(noquote("Error: Change length of coalescence block or sigma or log parameter")))
}



R/M.MSH.sampler.R
#' @include LR.R multishift.R
#'
# Function: M.MSH.sampler
#
# Metropolis with multishift coupler
# LB: time of block for coelescing
# init: initial values of two paths (the most interest range)
# post: log of target or postorior distribution
# sigma: standard deviation for multshift proposal
# log: A logical value of T or F which allows to log form of posterior
#
# Authors: Majid Nabipoor, Duncan Murdoch
# revised: Oct. 2021

M.MSH.sampler<- function(LB,init, post,sigma, log=FALSE)
{
  nn<- length(init)
  d<- matrix(rep(0,nn*(LB+1)), nrow=LB+1)
  d[1,]<- init
  for(i in 1:LB)
  {
    y<-multishift(d[i,],sigma)
    u<- runif(1,0,1)
    ratio<- post(y)/post(d[i,])
    if(log==TRUE) {
                    u<-log(u)
                    ratio<-post(y)-post(d[i,])
                  }
    d[i+1,]<- ifelse(u<ratio,y,d[i,])
  }
  return(d)
}



R/ROCFTP.MMS.R
#' @include LR.R multishift.R M.MSH.sampler.R CFTP.R
NULL

#'  Perfect Sampling
#'
#'  \code{ROCFTP.MMS}, Read Once Coupling From The Past, with Metropolis-Multishift
#'  is used to generate a perfect sample for a given posterior density based on the
#'  two extreme starting paths, minimum and maximum of the most interest range of
#'  the posterior. It uses the monotone random operation of multishift coupler
#'  which allows to sandwich all of the state space in one point. It means both
#'  Markov Chains starting from the maximum and minimum will be coalesced.
#'  The generated sample is independent from the starting points. It is useful
#'  for mixture distributions too. The output of this function is a real value as an exact
#'  draw from the posterior distribution.
#'
#' @param LB defines the length of each block. The algorithm starts new blocks to
#' find a coalescence of the two Markov chains started from extreme points of the
#' most interest range. So, if it is too small the algorithm after 50 blocks
#' repetition will give error message, and if it is too large then it will
#' increase generation time.
#'
#'@param start is a vector of initial values or the two extreme points of the most interest
#' range of posterior.
#'
#' @param post is the posterior which is defined in the form of an R function.
#'
#' @param sigma is a real value for standard deviation of multishift coupler. Multishift coupler
#' is constructed based on normal density. If the posterior is a mixture
#' distribution or a multi-modal distribution, then sigma should be chosen in such
#' a way that Markov chains easily jumps between modes, and if the sigma is
#' chosen small; the Markov chains may trap in one mode and don't coalesce.
#'
#'@param log has the default value of FALSE for density function of posteriors.
#' If TRUE; then posterior should be defined in log form.
#'
#' @section References:
#' Nabipoor M, Murdoch D. (2010) ROCFTP With Metropolis-Multishift Coupler, Summer
#' Research, Department of Statistical and actuarial sciences, University of
#' Western Ontario
#' @examples
#' #Unimodal posterior
#' post<- function(x)
#' {
#'   dnorm(x, mean=30, sd=1)
#' }
#' start<- c(20,40)
#' LB=30
#' ROCFTP.MMS(LB,start, post,1)
#' #Generate n i.i.d. exact sample
#' RG <- function(i){ROCFTP.MMS(30, c(20,40), post, 1)}
#' n <- 5
#' mrtx <- matrix(1:n, ncol=1)
#' dat <- apply(mrtx,1,RG)
#' qqnorm(dat, pch = 16, frame = FALSE)
#' qqline(dat, col = "steelblue", lwd = 3)
#'
#' # multimodal posterior
#' post <- function(x)
#' {
#'  0.2*dnorm(x, mean=-5, sd=1)+0.2*dnorm(x, mean=5, sd=1)+0.6*dnorm(x, mean=15, sd=1)
#' }
#' start <- c(-15,25)
#' LB <- 116
#' sigma <- 3.5
#' ROCFTP.MMS(LB,start, post,sigma) #generates one exact sample
#' #Generate n i.i.d. exact sample
#' RG <- function(i){ROCFTP.MMS(116, c(-15,25), post, 3.5)}
#' n <- 5
#' mrtx <- matrix(1:n, ncol=1)
#' apply(mrtx,1,RG)
#'
#' #log form of posterior
#' post<- function(x) {dnorm(x, mean=30, sd=1, log=TRUE)}
#' start<- c(20,40)
#' LB=100
#' sigma=0.5
#' ROCFTP.MMS(LB,start, post, sigma, log=TRUE)
#' @author
#' Majid Nabipoor: nabipoor@@ualberta.ca
#' Duncan Murdoch: murdoch.duncan@@gmail.com
#' @importFrom stats dnorm rnorm runif
#' @import vctrs
#' @export

ROCFTP.MMS<- function(LB, start, post, sigma, log=FALSE)
{
  if (LB <= 0 | length(LB) > 1 | is.numeric(LB)==FALSE) {
    return(noquote("Error: LB should be an integer"))
  }
  if (length(start) != 2 | is.vector(start)==FALSE) {
    return(noquote("Error: The length of start vector should be two"))
  }
  if (sigma <= 0 | length(sigma) > 1 | is.numeric(sigma)==FALSE) {
    return(noquote("Error: sigma should be a positive number"))
  }
  allresults<- CFTP(LB,start, post,sigma, log)
  if(is.character(allresults)) {return(allresults)} else {
  allresults<- allresults[,(ncol(allresults)-length(start)+1):(ncol(allresults))]
  allresults<- cbind(rep(NA,nrow(allresults)), allresults)
  x<- allresults[nrow(allresults),2]
  k<- c(0, nrow(allresults))
  repeat{
    init<- c(x, start)
    samples<-M.MSH.sampler(LB,init, post, sigma, log)
    y<-x
    x<- samples[nrow(samples),1]
    z<-matrix(c(rep(NA, (nrow(allresults))*length(init)-length(init)),init), ncol=ncol(samples), byrow=TRUE)
    z1<-matrix(rep(NA, (ncol(allresults))*nrow(samples)), ncol=ncol(allresults), byrow=TRUE)
    allresults<- cbind(allresults,z)
    sample<- cbind(z1,samples)
    allresults<- rbind(allresults,sample)
    k<- c(k, nrow(allresults))
    if(length(unique(samples[nrow(samples),]))==1) break
  }
  return(y) }
  if (is.numeric(y)==FALSE) {
    return(noquote("Error: No return value, check parameters"))
  }
}

